### PR TITLE
Support of custom events and event factories: fix #97

### LINKED ISSUE #97: allow to override factories in com.github.messenger4j.webhook.factory.EventFactory#FACTORIES
Hello @maxgrabenhorst , can you please allow us to supply custom factories in `com.github.messenger4j.webhook.factory.EventFactory#FACTORIES` ?

This will allow to supply custom processor for not-yet-implemented messages (for instance https://github.com/messenger4j/messenger4j/issues/95 - `template` attachments), and possible future API upgrades.

### PR DESCRIPTION
**Pull request by PhantomYdn**

### CODE REVIEW

File: src/main/java/com/github/messenger4j/webhook/factory/EventFactory.java [hunk 1]
```diff
@@ -20,25 +20,47 @@
...
+    	resetFactoriesDefault();
+    }
 
     private EventFactory() {
     }
+    
+    public static synchronized void regiterBaseEventFactory(BaseEventFactory<?> eventFactory) {
+		FACTORIES.add(eventFactory);
+    }
+
```
Comment (alexander-tampier) [inline src/main/java/com/github/messenger4j/webhook/factory/EventFactory.java]: Why don't you use for-each loop?

```java
for(BaseEventFactory baseEventFactory: eventFactories){
registerBaseEventFactory(baseEventFactory)
}
```

File: src/test/java/com/github/messenger4j/test/integration/DocumentationTest.java [hunk 2]
```diff
@@ -203,4 +215,67 @@ public void echoExample() throws Exception {
...
+    	
+    }
+    
+    public static class CustomEventFactory implements BaseEventFactory<CustomEvent> {
+
+		@Override
+		public boolean isResponsible(JsonObject messagingEvent) {
+			return messagingEvent.has("custom");
+		}
+
```
Comment (alexander-tampier) [inline src/test/java/com/github/messenger4j/test/integration/DocumentationTest.java]: Max used `Jsonhelper.getPropertyAsString(messagingEvent, CUSTOM).orEleseThrow(IllegalArgumentException::new)` for extracting a key out of JSON 👍

File: src/test/java/com/github/messenger4j/test/integration/DocumentationTest.java [hunk 3]
```diff
@@ -203,4 +215,67 @@ public void echoExample() throws Exception {
...
+			return new CustomEvent(messagingEvent.get("custom").getAsString());
+		}
+    	
+    }
+    
+    @Test
+    public void shouldAllowCustomEventFactory() throws Exception{
+        final Messenger messenger = Messenger.create("test", "test", "test");
+        assertThat(messenger, is(notNullValue()));
+
```
Comment (alexander-tampier) [inline src/test/java/com/github/messenger4j/test/integration/DocumentationTest.java]: Why is it necessary to support such an Event? Is this a Webhook event from facebook for messaging?

**Comment by maxgrabenhorst** [comment]
Thanks Ilia for your efforts implementing a generic mechanism for adding new event types. Could you please elaborate the uses cases in more detail? From my point of view there are no user specific event types at all. It is more like that we should focus on adding new event types faster to messenger4j, which are then available to all users. What do you think?

**Comment by maxgrabenhorst** [comment]
And thanks @alexander-tampier for your review!

**Comment by maxgrabenhorst** [comment]
Closed due to no response. Please feel free to reopen.

**Pull request closed without merging**